messaginginapp/src/main/java/io/customer/messaginginapp/ModuleMessagingInApp.kt

```diff
@@ -0,0 +1,74 @@
+package io.customer.messaginginapp
+
+import android.app.Application
+import io.customer.messaginginapp.di.gistProvider
+import io.customer.sdk.CustomerIO
+import io.customer.sdk.CustomerIOModule
+import io.customer.sdk.data.request.MetricEvent
+import io.customer.sdk.di.CustomerIOComponent
+import io.customer.sdk.hooks.HooksManager
+import io.customer.sdk.hooks.ModuleHook
+import io.customer.sdk.hooks.ModuleHookProvider
+import io.customer.sdk.repository.TrackRepository
+
+class ModuleMessagingInApp internal constructor(
+    private val overrideDiGraph: CustomerIOComponent?,
+    private val organizationId: String
+) : CustomerIOModule {
+
+    constructor(organizationId: String) : this(
+        overrideDiGraph = null,
+        organizationId = organizationId
+    )
+
+    override val moduleName: String
+        get() = "MessagingInApp"
+
+    private val diGraph: CustomerIOComponent
+        get() = overrideDiGraph ?: CustomerIO.instance().diGraph
+
+    private val trackRepository: TrackRepository
+        get() = diGraph.trackRepository
+
+    private val hooksManager: HooksManager by lazy { diGraph.hooksManager }
+
+    private val gistProvider by lazy { diGraph.gistProvider }
+
+    override fun initialize() {
+        initializeGist(organizationId)
+        setupHooks()
+        setupGistCallbacks()
+    }
+
+    private fun setupGistCallbacks() {
+        gistProvider.subscribeToEvents { deliveryID ->
+            trackRepository.trackInAppMetric(
+                deliveryID = deliveryID,
+                event = MetricEvent.opened
+            )
+        }
+    }
+
+    private fun setupHooks() {
+        hooksManager.add(object : ModuleHookProvider() {
+            override fun profileIdentifiedHook(hook: ModuleHook.ProfileIdentifiedHook) {
+                gistProvider.setUserToken(hook.identifier)
+            }
```

Comment:
🤔 will this cause memory issues in the future? 

`object : ModuleHookProvider()` creates a instance that is stored in `HooksManager` > `gistProvider` inside of the `ModuleHookProvider` is holding a strong reference to the `gistProvider` instance inside of the `ModuleMessagingInApp` class. This means that the `ModuleHookProvider` instance holds a strong reference to `GistProvider` and to `ModuleMessagingInApp`. 

I was expecting to see less strong references to instances. 

With the current implementation of `ModuleMessagingInApp`, this might not be a problem today. However, having these strong references here might cause issues in the future if we are not careful. 

Would it work if we had `ModuleHookProvider` create *new* instances of `GistProvider` on each function call? 

```kotlin
hooksManager.add(object : ModuleHookProvider() {
  private val diGraph: CustomerIOComponent 
    get() = CustomerIO.instance().diGraph

  val profileIdentifiedHook: ModuleHook.ProfileIdentifiedHook {
    return diGraph.gistProvider
  }
```

I like how an implementation like this doesn't have any strong references to anything. It's lightweight and less error-prone.

Reply:
While the concern is valid, I thought about the implementation where we are creating new instances as well, but I, later on, reached the conclusion that it shouldn't cause memory leakages. 

Since the SDK is `singleton` and the hooksManager is `singleton` too and all of its scopes is tied to the Application, it would all remain in memory anyways. 

```
modules: MutableMap<String, CustomerIOModule> = mutableMapOf()
```
inside `CustomerIO` is also going to be holding the same strong references for `ModuleMessagingInApp` and `ModuleMessagingPush`.

>this might not be a problem today However, having these strong references here might cause issues in the future if we are not careful.

I agree we will need to be careful on how we add on top of it, but I think with SDK being singleton we do have a certain blank of safety here and when we later work on, we can even look into keeping `WeakReference` both in `CustomerIO` class and `HooksManager`

Reply:
I do understand where you're coming from. I appreciate the thought you put into the implementation as you were writing it. 

It looks like you and I both agree that we do not see a problem with the code as it is today and we both agree that we need to be careful on how we add on top of this code in the future. 

With that being said, I want to ask:
1. Can we come up with a solution right now that removes the error prone code and so we no longer *need to be careful*? Is this new solution better then the current implementation in this PR? Can the new solution be implemented now before merging if it doesn't take too much time? 
2. If we can't do ^^^, create a ticket for this and work on it after we get the SDK deployed. This SDK is important to get out the door but I do think that the code in this implementation is important to improve if we come up with something better. Along with creating a ticket to improve the code, I think it would also be a good idea to add a comment to the code saying that it's code to be careful with to warn future developers touching that code until we come up with something better. 

I like code that we don't need to be careful using. We are only human and if we have code we know we need to be careful with, it's easy to make mistakes with it. I believe it's important to document careful code and try to improve it as a responsible way to address careful code.

Reply:
Thank you for bringing this up, I agree this needs a thorough discussion so it doesn't remain a repetitive topic. 

So, I'll try to target each comment. 

>Can we come up with a solution right now that removes the error-prone code so we no longer need to be careful? 

What is error-prone code over here, it's the fact that our `Modules`, i.e,  ModuleMessagingInApp and ModuleMessagingPush are going to be referenced by the `tracking` module, either in the form of `modules` in `CustomerIO` class or by `HooksManager`. 

>we do not see a problem with the code as it is today

We both agree, for now, it's not much of a problem because we have managed to keep `Modules` lightweight, and because of the fact SDK is a singleton, so only one instance is created and it's scoped by the Application lifecycle and even the modules are sharing the same `digraph`. 


>Can we come up with a solution right now that removes the error-prone code so we no longer need to be careful?

From what I can think of right now, whatever we have right now is the best way possible and it will stay that way until we don't bloat our `Modules`. For example, for all eventing in the modules, we just utilize the same `repositories`. If we were to start creating new classes and references that will then go on to be referenced by the tracking module, that's where the problems will start arising. 

If you have other ideas to improve, please share them here, keeping in mind the current implementation of iOS and Android is different in a sense, iOS does create and hold multiple instances of SDK, whereas Android doesn't. 

> think it would also be a good idea to add a comment to the code saying that it's code to be careful with to warn future developers touching that code until we come up with something better.

I would suggest, mentioning it on Modules to keep them light and to rely on repositories and mutually shared digraph 
for common code. Let me know, if this much sounds good or what would you add.

Reply:
In my previous comment where I say, "error-prone code" and "code to be careful using", I am not referring to 1 specific line of code. I am referring to the whole implementation of module hooks in this pull request. Because we have identified that we need to be careful when using the current implementation of module hooks, I am calling the entire implementation error-prone. 

I am suggesting that we either change the implementation of module hooks in this pull request to code that we no longer need to be careful using it, or, we create a ticket to come up with a solution in the future that we no longer need to be careful using. 

When I see code that we need to be careful with, I don't like to merge it without doing something about it first. If we were to merge the code as it is today without improving the implementation or at a minimum create a ticket for it, we are setting our team up for creating bugs in the future. I am trying to avoid future bugs from happening now before merging the code in. 

---

Looking at the implementation of the module hooks again in this class:
```kotlin
hooksManager.add(object : ModuleHookProvider() {
            override fun profileIdentifiedHook(hook: ModuleHook.ProfileIdentifiedHook) {
                gistProvider.setUserToken(hook.identifier)
            }
```

Besides talking about memory resources, I don't find this implementation scalable. At this time, each of the `ModuleHookProvider` functions implemented only call 1 dependency: `gistProvider`. But in the future, this could increase to many more dependencies making this file more complex. 

I would be OK if we move `object : ModuleHookProvider()` out of this class and into it's own class. Something like:
```kotlin
class ModuleInAppHookProvider: ModuleHookProvider() {

  private var diGraph: CustomerIOComponent {
    CustomerIO.instance().diGraph
  }

  private var gistProvider: GistProvider {
    diGraph.gistProvider
  }

  override fun profileIdentifiedHook(hook: ModuleHook.ProfileIdentifiedHook) {
                gistProvider.setUserToken(hook.identifier)
            }
```

I like how an implementation like this is simple but scales better then having everything in 1 file. Also, it removes the parts of the code that we keep saying "we need to be careful using" as we are getting dependencies directly out of the di graph here and keeping strong references to code to a minimum.

Reply:
>I am referring to the whole implementation of module hooks in this pull request. Because we have identified that we need to be careful when using the current implementation of module hooks, I am calling the entire implementation error-prone.

Even if we remove the `implementation of module hooks`, we will still need to be careful. Because in the whole `module implementation` initialization is done for `FCM token`, the `CustomerIO` class holds a strong reference of each `module` class through it. Now that would become error-prone as well?. So in any case, we need always make sure our modules are lighter. 

So regardless of any implementation, whenever multi-module need to communicate with each other, the base module would need to hold a reference. Having said that, as mentioned earlier, since our SDK is a singleton, we have the luxury of being in memory and all references are pointed to the same instance. 

> Besides talking about memory resources, I don't find this implementation scalable. At this time, each of the ModuleHookProvider functions implemented only call 1 dependency: gistProvider. But in the future, this could increase to many more dependencies making this file more complex.

Not sure what you mean by that, why would there be more dependencies in `ModuleMessagingInApp` other than gist or when we replace it with ours? and even if there are let's say even 2, not sure how that would make it complex, cause it would be just another line of code added here, and the provider is going to handle its own implementation in its own class. 

>I would be OK if we move object : ModuleHookProvider() out of this class and into it's own class.

Sure, I can do that, but as mentioned earlier, "we `still` need to be careful using" modules because its still being strongly referenced by `tracking` module.

Reply:
> Sure, I can do that, but as mentioned earlier, "we still need to be careful using" modules because its still being strongly referenced by tracking module.

I agree that in regards to modules, we do still need to be careful. I think it would be a good idea to add a comment to [the module interface](https://github.com/customerio/customerio-android/blob/213165333b947eedb8506f3bf4c3020dfb88305f/sdk/src/main/java/io/customer/sdk/CustomerIOModule.kt#L3-L7=) reminding the reader to keep modules lightweight since they are strongly referenced by the `tracking` module. Maybe there is a better place for a comment but I think a reminder would be helpful to avoid issues in the future. 

---

I see the [commit](https://github.com/customerio/customerio-android/pull/106/commits/213165333b947eedb8506f3bf4c3020dfb88305f) made to this PR. Looks great! I'll approve of this PR as it looks like feedback as all been resolved.